dimod/constrained.py

```diff
@@ -0,0 +1,395 @@
+
+        # ok, everything checks out so let's add it
+        for v in bqm.variables:
+            variables._append(vartype, v, permissive=True)
+
+        self._objective = bqm
+
+    def to_file(self, *, spool_size: int = int(1e9)) -> tempfile.SpooledTemporaryFile:
+        """Serialize to a file-like object.
+
+        Args:
+            spool_size: Defines the `max_size` passed to the constructor of
+                :class:`tempfile.SpooledTemporaryFile`. Determines whether
+                the returned file-like's contents will be kept on disk or in
+                memory.
+
+        Format Specification (Version 1.0):
+
+            This format is inspired by the `NPY format`_
+
+            The first 8 bytes are a magic string: exactly "DIMODCQM".
+
+            The next 1 byte is an unsigned byte: the major version of the file
+            format.
+
+            The next 1 byte is an unsigned byte: the minor version of the file
+            format.
+
+            The next 4 bytes form a little-endian unsigned int, the length of
+            the header data HEADER_LEN.
+
+            The next HEADER_LEN bytes form the header data. This is a
+            json-serialized dictionary. The dictionary is exactly:
+
+            .. code-block:: python
+
+                dict(num_variables=len(cqm.variables),
+                     num_constraints=len(cqm.constraints),
+                     num_biases=cqm.num_biases(),
+                     )
+
+            it is terminated by a newline character and padded with spaces to
+            make the entire length of the entire header divisible by 64.
+
+            The constraint quadratic model data comes after the header. It is
+            encoded as a zip file. The zip file will contain one file
+            named `objective`, containing the objective as encoded as a file
+            view. It will also contain a directory called `constraints`. The
+            `constraints` directory will contain one subdirectory for each
+            constraint, each containing `lhs`, `rhs` and `sense` encoding
+            the `lhs` as a fileview, the `rhs` as a float and the sense
+            as a string.
+
+        .. _NPY format: https://numpy.org/doc/stable/reference/generated/numpy.lib.format.html
+
+        """
+        file = SpooledTemporaryFile(max_size=spool_size)
+
+        # apply header
+        prefix = CQM_MAGIC_PREFIX
+        file.write(prefix)
+        file.write(bytes((1, 0)))  # version 1.0
+
+        data = dict(num_variables=len(self.variables),
+                    num_constraints=len(self.constraints),
+                    num_biases=self.num_biases(),
+                    )
+
+        header_data = json.dumps(data, sort_keys=True).encode('ascii')
+        header_data += b'\n'
+        header_data += b' '*(64 - (len(prefix) + 6 + len(header_data)) % 64)
+        header_len = np.dtype('<u4').type(len(header_data)).tobytes()
```

Comment:
I did just a quick check, but it looks `.tobytes()` in this setup actually always uses **native** byteorder. I believe your intention was to use little endian.

Reply:
Yup, excellent catch
```
In [6]: np.frombuffer(np.dtype('<u4').type(256).tobytes(), '<u4')
Out[6]: array([256], dtype=uint32)

In [7]: np.frombuffer(np.dtype('>u4').type(256).tobytes(), '<u4')
Out[7]: array([256], dtype=uint32)

In [8]: np.frombuffer(np.dtype('>u4').type(256).tobytes(), '>u4')
Out[8]: array([65536], dtype=uint32)

In [9]: np.frombuffer(np.dtype('<u4').type(256).tobytes(), '>u4')
Out[9]: array([65536], dtype=uint32)
```

Reply:
I see we have the same bug in BQM and DQM.

Reply:
Indeed. Investigating

Reply:
Have a fix I'll make in a followup PR.